File: 1- Data Structures/1- Basic/2- Stacks/stack.js
class Stack {
  constructor() {
      this.items = [];
  }

  // Adicionar um elemento ao topo da pilha
  push(element) {
      this.items.push(element);
  }

  // Remover o elemento do topo da pilha
  pop() {
      if (this.isEmpty()) {
          return "A pilha está vazia";
      }
      return this.items.pop();
  }

  // Ver o elemento do topo da pilha sem remover
  peek() {
      if (this.isEmpty()) {
          return "A pilha está vazia";
      }
      return this.items[this.items.length - 1];
  }

  // Verificar se a pilha está vazia
  isEmpty() {
      return this.items.length === 0;
  }

  // Ver o tamanho da pilha
  size() {
      return this.items.length;
  }

  // Esvaziar a pilha
  clear() {
      this.items = [];
  }
}

// Exemplo de utilização
const stack = new Stack();
stack.push(10);
stack.push(20);
stack.push(30);

console.log("Topo da pilha:", stack.peek());  // Output: 30
console.log("Removendo topo:", stack.pop());  // Output: 30
console.log("Topo da pilha:", stack.peek());  // Output: 20
console.log("Tamanho da pilha:", stack.size()); // Output: 2

File: 1- Data Structures/1- Basic/2- Stacks/stack.cpp
#include <iostream>
#include <stack>

using namespace std;

int main() {
    // Criando uma pilha (stack) de inteiros
    stack<int> s;

    // Adicionando elementos à pilha
    s.push(10);
    s.push(20);
    s.push(30);

    // Exibindo o elemento no topo da pilha
    cout << "Topo da pilha: " << s.top() << endl;  // Output: 30

    // Removendo o elemento do topo da pilha
    s.pop();
    cout << "Removendo o topo, novo topo: " << s.top() << endl;  // Output: 20

    // Exibindo o tamanho da pilha
    cout << "Tamanho da pilha: " << s.size() << endl;  // Output: 2

    // Verificando se a pilha está vazia
    if (s.empty()) {
        cout << "A pilha está vazia" << endl;
    } else {
        cout << "A pilha não está vazia" << endl;  // Output: A pilha não está vazia
    }

    return 0;
}
